PowerShell Scanners/Chromium-based Browser Extensions/Chromium-based Browser Extensions.ps1

```diff
@@ -0,0 +1,182 @@
+        'ProfileNames' = 'Opera*'
+        'Settings'     = 'opsettings'
+    }
+    'Vivaldi'        = @{
+        'AppData'      = 'Local'
+        'LastVersion'  = 'last_chrome_version'
+        'ProfileBase'  = 'Vivaldi\User Data'
+        'ProfileNames' = 'Default', 'Profile*'
+        'Settings'     = 'settings'
+    }
+}
+
+# Set up or check the list of browsers to scan.
+if ( -not $Browsers ) {
+
+    $Browsers = $BrowserTable.Keys
+
+} else {
+
+    Foreach ( $BrowserName in $Browsers ) {
+
+        if ( $BrowserName -notin $BrowserTable.Keys ) {
+
+            throw "'$BrowserName' does not match any entries in the list of supported browsers."
+
+        }
+
+    }
+
+}
+
+# Set up the JSON parser for the Preferences files below.
+# This .NET method is necessary because ConvertFrom-Json can't handle duplicate entries with different cases.
+# https://github.com/pdq/PowerShell-Scanners/issues/23
+Add-Type -AssemblyName System.Web.Extensions
+$JsonParser = New-Object -TypeName System.Web.Script.Serialization.JavaScriptSerializer
+
+Foreach ( $User in (Get-ChildItem -Directory -Path "$env:SystemDrive\Users") ) {
+
+    Foreach ( $BrowserName in $Browsers ) {
+
+        $Browser = $BrowserTable.$BrowserName
+    
+        # Get profiles.
+        $ProfileBase = "$($User.FullName)\AppData\$($Browser.AppData)\$($Browser.ProfileBase)"
+        if ( Test-Path $ProfileBase ) {
+
+            Set-Location -Path $ProfileBase
+
+        } else {
+
+            # Browser is not installed, or the user has never opened it.
+            Continue
+
+        }
+        $Profiles = Get-Item -Path $Browser.ProfileNames
+
+        Foreach ( $Profile in $Profiles ) {
+        
+            $SecurePreferencesFile = "$($Profile.FullName)\Secure Preferences"
+            if ( Test-Path $SecurePreferencesFile ) {
+
+                # Secure Preferences contains the vast majority of the information we want.
+                $SecurePreferencesText = Get-Content $SecurePreferencesFile
+                $SecurePreferencesJson = $JsonParser.DeserializeObject($SecurePreferencesText)
+
+            } else {
+
+                Write-Verbose "Unable to find a 'Secure Preferences' file in: $($Profile.FullName)"
+                Continue
+
+            }
+
+            $PreferencesFile = "$($Profile.FullName)\Preferences"
+            $PreferencesJson = $null
+            if ( Test-Path $PreferencesFile ) {
+
+                # The only thing we care about in Preferences is the last browser version.
+                $PreferencesText = Get-Content $PreferencesFile
+                $PreferencesJson = $JsonParser.DeserializeObject($PreferencesText)
+
+            } else {
+
+                Write-Verbose "Unable to find a 'Preferences' file in: $($Profile.FullName)"
+                # It doesn't matter if this one is missing, so no need to skip.
+
+            }
+
+            Foreach ( $Extension in $SecurePreferencesJson.extensions."$($Browser.Settings)".GetEnumerator() ) {
```

Comment:
P.S. I posted this weeks ago, but my smart self didn't click the right button so it sat as "pending" until now...my bad.

Reply:
Ooh, nice catch, thanks! I just pushed a fix.

Reply:
Did your changes make it through? Git shows you deleted the previous commit and put in a new one, but I'm still seeing the same behavior.

Reply:
Yes, I rebased and force-pushed. I see the changes starting at line 97. What is the full error message you are getting?

Reply:
I took another look and I think I know what the problem is. I suspect that the Secure Preferences file on the troublesome target does not have an `extensions.settings` node. I removed my profile checking code because I think it is unecessary (the `Foreach` will skip if `$Profiles` is empty), and added a section to make sure `extensions.settings` exists and log a message if it doesn't.

Reply:
Not sure why I'm seeing this, but the scanner frequently errors out with: 
`Cannot find path 'C:\Users\*****\AppData\Local\Google\Chrome\User Data\Default' because it does not exist.
   At line:96 char:21`

Where `*****` is a mix of local and domain accounts.

Reply:
Ooh, good catch. `Get-Item` throws an error if you give it a regular name (such as 'Default') and it doesn't exist. However, it doesn't throw an error if you use a wildcard ('Default*'). This could potentially cause problems if any future folders start with 'Default', but the solution I came up with for that is much more involved. I'd like to use the simple solution for now :)

Reply:
It doesn't appear Google Chrome extensions are stored in the "Secure Preferences" location on my machine so no Chrome extensions were returned.  I'll have to dive deeper Monday and also check other browsers.  Just wanted to make sure I didn't forget.

Reply:
@PDQ-Zach I just pushed a change that should fix that issue.

Reply:
@CaseyM-PDQ Fixed, I believe. I don't have older systems to test on.

Reply:
@CaseyM-PDQ How are those older systems faring?